### PR TITLE
Handle overflows while calculating cost of the block in gas-metering algorithm

### PR DESCRIPTION
**Pull request by SabaunT**
A possibility of the overflow while calculating cost of the block in gas-metering algorithm depends on 2 factors: the cost of each instruction defined by the caller, i.e. user of the `wasm-instrument` crate, and the size of the wasm module (amount of instructions). Some projects can face overflow error while injecting gas calls, because of specific for them parameters of wasm module size or cost of each instruction. I believe handling overflows while cost (gas) calculation can give more freedom to `wasm-instrument` users.

The suggested in the PR solution:
1. Enable overflows counter while metering blocks
2. When injecting gas calls, push several gas call instructions instead of one as it is now.

### CODE REVIEW

File: src/gas_metering/mod.rs [hunk 1]
```diff
@@ -284,7 +284,71 @@ struct MeteredBlock {
...
+	fn costs_num(&self) -> usize {
+		// Block's total cost consists of accumulated overflows and the remained in accumulator
+		// value
+		(self.overflows + 1) as usize
+	}
+
+	/// Returns the tuple of costs, where the first element is an amount of overflows
+	/// emerged when summating block's cost, and the second element is the current
+	/// (not overflowed remainder) block's cost.
+	fn block_costs(&self) -> (u32, u32) {
```
Comment (NikVolf) [inline src/gas_metering/mod.rs:349]: why not just return all costs?

Comment (SabaunT) [reply src/gas_metering/mod.rs:349]: This is done for optimisation reasons in order not to allocate redundant memory. 
An amount of overflows is dynamic, so to return all costs we need to return them in some container like vector. We can avoid this by returning the "length of the vector" which is the first parameter of the returning tuple.

**Comment by athei** [comment]
I think we should opt for a simpler approach: Change the counter to `u64` and just error out on overflow. Don't know why `u32` was chosen but `u64` does not incur any overhead as an primitive.

**Review by athei — CHANGES_REQUESTED** [review]
I think we should opt for a simpler approach: Change the counter to u64 and just error out on overflow. Don't know why u32 was chosen but u64 does not incur any overhead as an primitive. Of course VMs need to add a new imported function but that shouldn't be a problem.

**Comment by NikVolf** [comment]
> I think we should opt for a simpler approach: Change the counter to u64 and just error out on overflow. Don't know why u32 was chosen but u64 does not incur any overhead as an primitive. Of course VMs need to add a new imported function but that shouldn't be a problem.

Embedders (multiple blockchain projects) already have host functions with u32 everywhere, they will hardly change it.
This is very desirable fix without changing signature.

**Comment by athei** [comment]
Why? They will need to update their runtime to deploy this change. The same update adds the new exported function then. Why will they change one thing but not the other.

**Comment by NikVolf** [comment]
> Why? They will need to update their runtime to deploy this change. The same update adds the new exported function then. Why will they change one thing but not the other.

This code is used not only in substrate where you can update runtime.

And also, we can merge this fix for u32, bump to 0.2.1 and then rewrite to u64 and bump to 0.3.0. I believe changing from u32 to u64 for generated function signature should be a major version bump despite formally only implementation details of this library changed.

**Review by NikVolf — APPROVED** [review]

**Comment by SabaunT** [comment]
> Why? They will need to update their runtime to deploy this change. The same update adds the new exported function then. Why will they change one thing but not the other.

I totally agree with @NikVolf 's suggested solution with versioning changes. One more pros for applying these changes is giving more freedom to crate users in defining arbitrarily large costs for instructions on their side. `u64` doesn't save from overflow, but this [677f8fe](https://github.com/paritytech/wasm-instrument/pull/29/commits/677f8fe95999a9e9ae4cc323bed4f5672114d70a) does.

**Comment by athei** [comment]
> This code is used not only in substrate where you can update runtime.

Yes but I need to carry the maintenance burden here and stuff that is not needed for substrate I will not maintain. Please fork in that case.

>  I believe changing from u32 to u64 for generated function signature should be a major version bump despite formally only implementation details of this library changed.

I agree. Never said anything else.

> u64 doesn't save from overflow

`u64` + error on overflow. We simply do not support basic blocks bigger than `u64` then.

**Comment by NikVolf** [comment]
I don't suggest you maintain anything

We patch to 0.2.1 by fixing this bug, publish
Then replace with u64 and publish 0.3

**Comment by NikVolf** [comment]
Or you can just publish 0.2.1 from the branch @athei

**Comment by athei** [comment]
Why not just release it as another crate and use this? No one forces you to use this crate. I think this is way too much complexity for something that can just be put to `u64` and call it a day. Remember that this counter is per basic block not for the whole contract.

**Pull request closed without merging**